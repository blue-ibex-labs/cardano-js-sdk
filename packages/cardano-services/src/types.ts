export type ModuleState = null | 'initializing' | 'initialized';
export type Awaited<T> = T extends PromiseLike<infer U> ? U : T;
