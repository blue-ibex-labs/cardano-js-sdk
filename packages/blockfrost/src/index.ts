export { WalletProvider } from '@cardano-sdk/core';
export * from './blockfrostWalletProvider';
export * from './blockfrostAssetProvider';
export * from './blockfrostTxSubmitProvider';
export * from './blockfrostNetworkInfoProvider';
export * from './blockfrostUtxoProvider';
export { Options } from '@blockfrost/blockfrost-js/lib/types';
export { BlockFrostAPI } from '@blockfrost/blockfrost-js';
