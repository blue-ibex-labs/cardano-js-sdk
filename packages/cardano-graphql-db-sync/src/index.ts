export { WalletProvider } from '@cardano-sdk/core';
export * from './cardanoGraphqlDbSyncProvider';
export * from './cardanoGraphqlDbSyncTxSubmitProvider';
