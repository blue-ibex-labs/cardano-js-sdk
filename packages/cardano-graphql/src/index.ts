export { WalletProvider, StakePoolSearchProvider } from '@cardano-sdk/core';
export * from './StakePoolSearchProvider';
export * from './WalletProvider';
export * from './AssetProvider';
export * from './createSDK';
// Do not export this. Using core types elsewhere in the sdk.
// export * as Schema from './Schema';
