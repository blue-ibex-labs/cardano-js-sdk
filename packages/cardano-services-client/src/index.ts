export { TxSubmitProvider } from '@cardano-sdk/core';
export * from './HttpProvider';
export * from './TxSubmitProvider';
export * from './StakePoolSearchProvider';
export * from './UtxoProvider';
