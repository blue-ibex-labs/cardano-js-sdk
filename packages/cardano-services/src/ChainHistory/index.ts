export * from './ChainHistoryHttpService';
export * from './DbSyncChainHistoryProvider';
