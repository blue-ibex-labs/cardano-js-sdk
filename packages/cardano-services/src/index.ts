export * from './Http';
export * from './Program';
export * from './RunnableModule';
export * from './StakePoolSearch';
export * from './TxSubmit';
export * from './Utxo';
export * as errors from './errors';
