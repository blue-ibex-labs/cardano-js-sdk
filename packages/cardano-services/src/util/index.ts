export * from './http';
export * from './logging';
export * from './provider';
export * from './hexString';
