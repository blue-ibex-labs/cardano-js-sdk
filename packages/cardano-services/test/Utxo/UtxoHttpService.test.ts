/* eslint-disable max-len */
import { Cardano, UtxoProvider } from '@cardano-sdk/core';
import { DbSyncUtxoProvider, HttpServer, HttpServerConfig, UtxoHttpService } from '../../src';
import { Pool } from 'pg';
import { getPort } from 'get-port-please';
import { utxoHttpProvider } from '@cardano-sdk/cardano-services-client';
import got from 'got';

const APPLICATION_JSON = 'application/json';
const APPLICATION_CBOR = 'application/cbor';
const UNSUPPORTED_MEDIA_STRING = 'Response code 415 (Unsupported Media Type)';
const BAD_REQUEST_STRING = 'Response code 400 (Bad Request)';

const toCardanoAddresses = (addresses: string[]) => addresses.map((a) => Cardano.Address(a));

describe('UtxoHttpService', () => {
  let dbConnection: Pool;
  let httpServer: HttpServer;
  let utxoProvider: DbSyncUtxoProvider;
  let service: UtxoHttpService;
  let port: number;
  let apiUrlBase: string;
  let config: HttpServerConfig;
  let provider: UtxoProvider;

  beforeAll(async () => {
    port = await getPort();
    apiUrlBase = `http://localhost:${port}/utxo`;
    config = { listen: { port } };
    dbConnection = new Pool({ connectionString: process.env.DB_CONNECTION_STRING });
  });

  afterEach(async () => {
    jest.resetAllMocks();
  });

  describe('healthy state', () => {
    beforeAll(async () => {
      utxoProvider = new DbSyncUtxoProvider(dbConnection);
      service = UtxoHttpService.create({ utxoProvider });
      httpServer = new HttpServer(config, { services: [service] });
      provider = utxoHttpProvider(apiUrlBase);
      await httpServer.initialize();
      await httpServer.start();
    });

    afterAll(async () => {
      await dbConnection.end();
      await httpServer.shutdown();
    });

    describe('/health', () => {
      it('/health response should be true', async () => {
        const res = await got.post(`${apiUrlBase}/health`, {
          headers: { 'Content-Type': APPLICATION_JSON }
        });
        expect(res.statusCode).toBe(200);
        expect(JSON.parse(res.body)).toEqual({ ok: true });
      });
      it('with utxoProvider', async () => {
        const response = await provider.healthCheck();
        expect(response).toEqual({ ok: true });
      });
    });
    describe('/search', () => {
      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await got.post(`${apiUrlBase}/search`, {
            headers: { 'Content-Type': APPLICATION_CBOR }
          });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.statusCode).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });
      it('returns 400 coded respons if the request is bad formed', async () => {
        try {
          await got.post(`${apiUrlBase}/search`, {
            json: { addresses: ['asd'] }
          });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.statusCode).toBe(400);
          expect(error.message).toBe(BAD_REQUEST_STRING);
        }
      });
      it('valid request should pass OpenApi schema validations', async () => {
        const req = ['asd'];
        const res = await got.post(`${apiUrlBase}/search`, {
          json: { args: [req] }
        });
        expect(res.statusCode).toEqual(200);
      });
      it('should search by a single address', async () => {
        const res = await utxoProvider.utxoByAddresses([
          Cardano.Address(
            'addr_test1qretqkqqvc4dax3482tpjdazrfl8exey274m3mzch3dv8lu476aeq3kd8q8splpsswcfmv4y370e8r76rc8lnnhte49qqyjmtc'
          )
        ]);
        expect(res).toMatchSnapshot();
      });
      it('should search by multiple addresses', async () => {
        const addresses = [
          'addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7',
          'addr_test1qr4m502gr9hnaxac5mxjln22jwavf7pcjmh9sw7fujdvgvj9ef6afquphwg7tj4mmm548m3t50hxfyygjuu222kx96eshcathg',
          'addr_test1qz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer3jcu5d8ps7zex2k2xt3uqxgjqnnj83ws8lhrn648jjxtwq2ytjqp'
        ];
        const res = await utxoProvider.utxoByAddresses(toCardanoAddresses(addresses));
        expect(res).toMatchSnapshot();
      });
      it('should successfully return utxos with multi assets', async () => {
        const addresses = [
          'addr_test1qrcj98ukemwfuwc72ad95yydnx83qch6s7plr8rg44nxv53fumt3ljeck26752eajzyavd8my3cp8cx3x2c538lx7h5swm4j4n',
          'addr_test1qretqkqqvc4dax3482tpjdazrfl8exey274m3mzch3dv8lu476aeq3kd8q8splpsswcfmv4y370e8r76rc8lnnhte49qqyjmtc'
        ];
        const res = await utxoProvider.utxoByAddresses(toCardanoAddresses(addresses));
        expect(res).toMatchSnapshot();
      });
    });
  });
});
