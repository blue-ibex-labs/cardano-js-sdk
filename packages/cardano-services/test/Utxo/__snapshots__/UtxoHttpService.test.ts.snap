// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UtxoHttpService healthy state /search should search by a single address 1`] = `Array []`;

exports[`UtxoHttpService healthy state /search should search by multiple addresses 1`] = `
Array [
  Array [
    Object {
      "address": "addr_test1qr4m502gr9hnaxac5mxjln22jwavf7pcjmh9sw7fujdvgvj9ef6afquphwg7tj4mmm548m3t50hxfyygjuu222kx96eshcathg",
      "index": 0,
      "txId": "8fd14baca91c674fafae59701b7dc0eda1266202ec8445bad3244bd8669a7fb5",
    },
    Object {
      "address": "addr_test1qr4m502gr9hnaxac5mxjln22jwavf7pcjmh9sw7fujdvgvj9ef6afquphwg7tj4mmm548m3t50hxfyygjuu222kx96eshcathg",
      "value": Object {
        "coins": Object {
          "__type": "bigint",
          "value": "999502622402",
        },
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer3jcu5d8ps7zex2k2xt3uqxgjqnnj83ws8lhrn648jjxtwq2ytjqp",
      "index": 0,
      "txId": "a62624fdc47e8b774ddff11a9a56cae5fa9b072975af87bf0a0583fca0e345f4",
    },
    Object {
      "address": "addr_test1qz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer3jcu5d8ps7zex2k2xt3uqxgjqnnj83ws8lhrn648jjxtwq2ytjqp",
      "value": Object {
        "coins": Object {
          "__type": "bigint",
          "value": "19717868703",
        },
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qr4m502gr9hnaxac5mxjln22jwavf7pcjmh9sw7fujdvgvj9ef6afquphwg7tj4mmm548m3t50hxfyygjuu222kx96eshcathg",
      "index": 0,
      "txId": "3d4cd09885d39673125c3a15f8acb45d92fde137f9effe7a5131f6cc7241d960",
    },
    Object {
      "address": "addr_test1qr4m502gr9hnaxac5mxjln22jwavf7pcjmh9sw7fujdvgvj9ef6afquphwg7tj4mmm548m3t50hxfyygjuu222kx96eshcathg",
      "value": Object {
        "coins": Object {
          "__type": "bigint",
          "value": "10306556787917",
        },
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
      "index": 0,
      "txId": "19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d",
    },
    Object {
      "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
      "value": Object {
        "coins": Object {
          "__type": "bigint",
          "value": "1497246388",
        },
      },
    },
  ],
]
`;

exports[`UtxoHttpService healthy state /search should successfully return utxos with multi assets 1`] = `
Array [
  Array [
    Object {
      "address": "addr_test1qretqkqqvc4dax3482tpjdazrfl8exey274m3mzch3dv8lu476aeq3kd8q8splpsswcfmv4y370e8r76rc8lnnhte49qqyjmtc",
      "index": 0,
      "txId": "573b137128145227be0f46da6408d8890ab265d238fc638f690950724eab0473",
    },
    Object {
      "address": "addr_test1qretqkqqvc4dax3482tpjdazrfl8exey274m3mzch3dv8lu476aeq3kd8q8splpsswcfmv4y370e8r76rc8lnnhte49qqyjmtc",
      "value": Object {
        "assets": Object {
          "__type": "Map",
          "value": Array [
            Array [
              "4eecf4013e2ab14ec8e500fafeafbdf2bd060254d76c9629a65b7f7054455354",
              Object {
                "__type": "bigint",
                "value": "1",
              },
            ],
            Array [
              "8615849a049172dcd729fc3185032b266ceeba4dc21a0e61e713ebf85445535432",
              Object {
                "__type": "bigint",
                "value": "1",
              },
            ],
          ],
        },
        "coins": Object {
          "__type": "bigint",
          "value": "2200000",
        },
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qrcj98ukemwfuwc72ad95yydnx83qch6s7plr8rg44nxv53fumt3ljeck26752eajzyavd8my3cp8cx3x2c538lx7h5swm4j4n",
      "index": 1,
      "txId": "573b137128145227be0f46da6408d8890ab265d238fc638f690950724eab0473",
    },
    Object {
      "address": "addr_test1qrcj98ukemwfuwc72ad95yydnx83qch6s7plr8rg44nxv53fumt3ljeck26752eajzyavd8my3cp8cx3x2c538lx7h5swm4j4n",
      "value": Object {
        "assets": Object {
          "__type": "Map",
          "value": Array [
            Array [
              "4eecf4013e2ab14ec8e500fafeafbdf2bd060254d76c9629a65b7f7054455354",
              Object {
                "__type": "bigint",
                "value": "998762",
              },
            ],
            Array [
              "8615849a049172dcd729fc3185032b266ceeba4dc21a0e61e713ebf85445535432",
              Object {
                "__type": "bigint",
                "value": "998762",
              },
            ],
          ],
        },
        "coins": Object {
          "__type": "bigint",
          "value": "100233833966",
        },
      },
    },
  ],
]
`;
